toolchain/sem_ir/ids.h

```diff
@@ -957,6 +957,25 @@ struct LocId : public IdBase, public Printable<LocId> {
 
 constexpr LocId LocId::Invalid = LocId(Parse::NodeId::Invalid);
 
+// Polymorphic id for fields in `Any[...]` typed instruction groups. Used for
+// fields where the specific instruction structs have different field types in
+// that position or do not have a field in that position at all. Allows
+// conversion with `Inst::As<>` from the specific typed instruction to the
+// `Any[...]` instruction group.
+//
+// This type participates in `Inst::FromRaw` in order to convert from specific
+// instructions to an `Any[...]` instruction group, but does not participate in
+// `Inst::ToRaw` as it's not possible to convert in the other direction:
```

Comment:
Historically we included these `int32_t` fields in order to support going in the other direction (for example, converting some branch instruction to `AnyBranch`, modifying the `target_id`, and then converting it back to an `Inst`). While it's definitely not type-safe in general to go in the other direction, I think it's not really much worse than instruction categories are in general in that regard -- there's no enforcement that the types of fields in the category match the types of fields in the individual instructions at the moment in general.

We do perform conversions in the category -> instruction direction in some cases ([example](https://github.com/carbon-language/carbon-lang/blob/74dcd1fd05771b1963d12d10c74db6e6ec70a6e6/toolchain/check/eval.cpp#L500)), but it looks like we don't do so for any instruction categories with an "untyped" `int32_t` field at the moment. Given that we don't use this functionality for non-homogeneous instruction categories, I think I'd be OK with losing it, but it would be a little inconsistent I suppose.

Overall I think I don't have a strong opinion here, mostly just providing some context.

Reply:
I think it won't work in general right now, if the number of fields don't match. It might try to construct a missing second field from the AnyRawId, though I didn't test it out, but it could be made to work if not.

I think it's fair to allow both directions, and making it subclass IdBase means it will support both direction (as long as the conversion machinery handles the field number mismatch if the AnyRawId has no field to construct into). I removed that part of the comment.